File: python/src/sidebar.py
import os
import sys

sys.path.append('./src')

from application import Application  # noqa: E402


class Sidebar(Application):
    """Generates the wiki _Sidebar.md as a nested list of owners/categories
    and their wiki pages."""

    def __init__(self, base_path: str = '', group_by: str = '',
                 language: str = '',
                 home_overflow: str | bool = 'false') -> None:
        super().__init__(
            base_path=base_path, group_by=group_by,
            language=language, home_overflow=home_overflow,
        )
        self._base_owner_url = (
            f'https://github.com/orgs/'
            f'{os.environ.get("ORGANISATION_NAME", "hayat01sh1da")}/teams/'
        )
        self._wiki_list: list[str] = []

    # private

    def _run(self) -> None:
        self._update_wiki_list()
        with open(self._path_to_sidebar, 'w') as f:
            f.write(''.join(self._wiki_list))

    def _update_wiki_list(self) -> None:
        for namespace, wikis in self._owned_wiki_maps().items():
            self._append_section(namespace, wikis, owned=True)
        for namespace, wikis in self._plain_wiki_maps().items():
            self._append_section(namespace, wikis, owned=False)

    def _append_section(self, namespace: str, wikis: list[str],
                        owned: bool) -> None:
        self._wiki_list.append(
            f'{self._section_heading(namespace, owned=owned)}\n')
        for wiki in wikis:
            self._wiki_list.append(f'  - [[{wiki.replace(".md", "")}]]\n')

    def _section_heading(self, namespace: str, owned: bool) -> str:
        if owned:
            url = self._base_owner_url + namespace.replace('@', '')
            return f'- [{namespace}]({url})'
        return f'- {namespace}'

File: ruby/src/sidebar.rb
# frozen_string_literal: true
# rbs_inline: enabled

require_relative 'application'

# Generates the wiki _Sidebar.md as a nested list of owners/categories and
# their wiki pages.
class Sidebar < Application
  # @rbs base_path: String
  # @rbs group_by: String
  # @rbs language: String
  # @rbs home_overflow: String
  # @rbs array: Array[untyped]
  # @rbs return: void
  def initialize(base_path: '', group_by: '', language: '', home_overflow: 'false', array: [])
    super(base_path:, group_by:, language:, home_overflow:)
    @base_owner_url = "https://github.com/orgs/#{ENV.fetch('ORGANISATION_NAME', 'hayat01sh1da')}/teams/"
    @wiki_list      = array
  end

  # @rbs return: void
  def run
    update_wiki_list
    File.write(path_to_sidebar, wiki_list.join)
  end

  private

  attr_reader :base_owner_url, :wiki_list

  # @rbs return: void
  def update_wiki_list
    owned_wiki_maps.each { |namespace, wikis| append_section(namespace, wikis, owned: true) }
    plain_wiki_maps.each { |namespace, wikis| append_section(namespace, wikis, owned: false) }
  end

  # @rbs namespace: String
  # @rbs wikis: Array[String]
  # @rbs owned: bool
  # @rbs return: void
  def append_section(namespace, wikis, owned:)
    wiki_list << "#{section_heading(namespace, owned:)}\n"
    wikis.each { |wiki| wiki_list << "  - [[#{wiki.gsub('.md', '')}]]\n" }
  end

  # @rbs namespace: String
  # @rbs owned: bool
  # @rbs return: String
  def section_heading(namespace, owned:)
    owned ? "- [#{namespace}](#{base_owner_url + namespace.delete('@')})" : "- #{namespace}"
  end
end
